File: src/org/genyris/interp/AbstractEnvironment.java
// Copyright 2008 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.interp;

import org.genyris.core.DynamicSymbol;
import org.genyris.core.Exp;
import org.genyris.core.Internable;
import org.genyris.core.SimpleSymbol;
import org.genyris.core.Symbol;
import org.genyris.exception.GenyrisException;

public abstract class AbstractEnvironment implements Environment {

	public Exp lookupVariableValue(Symbol symbol) throws UnboundException {
		return symbol.lookupVariableValue(this);
	}

	public abstract Exp getSelf() throws UnboundException;

	public Exp lookupDynamicVariableValue(DynamicSymbol symbol)
			throws UnboundException {
		throw new UnboundException("no dynamic variable in environment: "
				+ symbol);
	}

	public abstract Exp lookupLexicalVariableValue(SimpleSymbol symbol)
			throws UnboundException;

	public void defineVariable(Symbol symbol, Exp valu) throws GenyrisException {
		if (symbol.isNil()) {
			throw new GenyrisException("Attempt to re-define nil!");
		}
		symbol.defineVariable(this, valu);
	}

	public void defineLexicalVariable(SimpleSymbol symbol, Exp valu)
			throws GenyrisException {
		throw new GenyrisException(
				"defineLexicalVariable: no simple variable in environment: "
						+ symbol);
	}

	public void defineDynamicVariable(DynamicSymbol symbol, Exp valu)
			throws GenyrisException {
		throw new GenyrisException(
				"defineDynamicVariable: no dynamic variable in environment: "
						+ symbol);
	}

	public abstract SimpleSymbol getNil();

	public abstract Internable getSymbolTable();

	public abstract Symbol internString(String symbolName);

	public abstract Exp lookupInThisClassAndSuperClasses(DynamicSymbol symbol)
			throws UnboundException;

	public void setVariableValue(Symbol symbol, Exp valu)
			throws UnboundException {
		if (symbol.isNil()) {
			throw new UnboundException("Attempt to re-define nil!");
		}
		symbol.setVariableValue(this, valu);
	}

	public void setLexicalVariableValue(SimpleSymbol symbol, Exp valu)
			throws UnboundException {
		throw new UnboundException(
				"setLexicalVariableValue: no dynamic variable in environment: "
						+ symbol);
	}

	public void setDynamicVariableValue(DynamicSymbol symbol, Exp valu)
			throws UnboundException {
		throw new UnboundException(
				"setDynamicVariableValue: no lexical variable in environment: "
						+ symbol);
	}
}

File: src/org/genyris/interp/AbstractEnvironment.kt
// Copyright 2008 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.interp

import org.genyris.core.*
import org.genyris.exception.GenyrisException

abstract class AbstractEnvironment : Environment {
    @kotlin.Throws(UnboundException::class)
    override fun lookupVariableValue(symbol: Symbol): Exp? {
        return symbol.lookupVariableValue(this)
    }

    @kotlin.Throws(UnboundException::class)
    abstract override fun getSelf(): Exp?

    @kotlin.Throws(UnboundException::class)
    override fun lookupDynamicVariableValue(symbol: DynamicSymbol?): Exp? {
        throw UnboundException(
            "no dynamic variable in environment: "
                    + symbol
        )
    }

    @kotlin.Throws(UnboundException::class)
    abstract override fun lookupLexicalVariableValue(symbol: SimpleSymbol?): Exp?

    @kotlin.Throws(GenyrisException::class)
    override fun defineVariable(symbol: Symbol, valu: Exp?) {
        if (symbol.isNil()) {
            throw GenyrisException("Attempt to re-define nil!")
        }
        symbol.defineVariable(this, valu)
    }

    @kotlin.Throws(GenyrisException::class)
    abstract override fun defineLexicalVariable(symbol: SimpleSymbol?, valu: Exp?)

    @kotlin.Throws(GenyrisException::class)
    override fun defineDynamicVariable(symbol: DynamicSymbol?, valu: Exp?) {
        throw GenyrisException(
            "defineDynamicVariable: no dynamic variable in environment: "
                    + symbol
        )
    }

    abstract override fun getNil(): SimpleSymbol?

    abstract override fun getSymbolTable(): Internable?

    abstract override fun internString(symbolName: String?): Symbol?

    @kotlin.Throws(UnboundException::class)
    abstract override fun lookupInThisClassAndSuperClasses(symbol: DynamicSymbol?): Exp?

    @kotlin.Throws(UnboundException::class)
    override fun setVariableValue(symbol: Symbol, valu: Exp?) {
        if (symbol.isNil()) {
            throw UnboundException("Attempt to set to nil!")
        }
        symbol.setVariableValue(this, valu)
    }

    @kotlin.Throws(UnboundException::class)
    abstract override fun setLexicalVariableValue(symbol: SimpleSymbol?, valu: Exp?)

    @kotlin.Throws(UnboundException::class)
    override fun setDynamicVariableValue(symbol: DynamicSymbol?, valu: Exp?) {
        throw UnboundException(
            "setDynamicVariableValue: no lexical variable in environment: "
                    + symbol
        )
    }
}